File: Example/PayTPV Objc Example/ProductsTableViewController.m
//
//  ProductsTableViewController.m
//  PAYTPV Objc Example
//
//  Created by Mihail Cristian Dumitru on 8/18/17.
//  Copyright © 2017 PAYTPV. All rights reserved.
//

#import "PAYTPV/PAYTPV.h"

#import "ProductsTableViewController.h"
#import "CardsTableViewController.h"
#import "PurchasesViewController.h"

#import "Models.h"
#import "AlertBuilder.h"
#import "ModelStore.h"

@interface ProductsTableViewController () <CardsTableControllerDelegate>

#pragma mark - Properties

@property (nonatomic) NSArray *products;
@property (nonatomic) Product *selectedProduct;

@end

@implementation ProductsTableViewController

#pragma mark - Lifecycle

- (void)viewDidLoad {
    [super viewDidLoad];

    [self setup];
}

- (void)setup {
    self.title = @"Products";

    self.products = @[[[Product alloc] initWithName:@"Product A" price:@0.29 currency:PTPVCurrencyEUR],
                      [[Product alloc] initWithName:@"Product B" price:@1.99 currency:PTPVCurrencyEUR],
                      [[Product alloc] initWithName:@"Product C" price:@2.99 currency:PTPVCurrencyEUR],
                      [[Product alloc] initWithName:@"Product D" price:@3.99 currency:PTPVCurrencyEUR]
                      ];

    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Purchases"
                                                                             style:UIBarButtonItemStylePlain
                                                                            target:self action:@selector(onPurchasesButtonPressed)];
}

#pragma mark - Helpers

- (void)onPurchasesButtonPressed {
    [self.navigationController pushViewController:[PurchasesViewController new] animated:true];
}
- (void)purchaseProduct:(Product *)product
               withCard:(Card *)card {
    // add a loading indicator
    AlertBuilder *loader = [AlertBuilder new];
    [loader setTitle:@"Executing purchase"];
    [loader addLoadingIndicator];
    [loader presentInController:self animated:true completion:nil];

    // the amount must be in integer format, so we have to convert it
    // e.g. 1.00 EURO = 100, 4.50 EUROS = 450...
    NSNumber *amount = [NSNumber numberWithLong:lround([[product price] doubleValue] * 100.0)];

    // generate unique order reference
    NSString *order = [NSString stringWithFormat:@"%.0f_%d", [[NSDate date] timeIntervalSince1970], arc4random_uniform(1000)];

    // create the purchase request
    PTPVPurchaseRequest *purchaseRequest;
    purchaseRequest = [[PTPVPurchaseRequest alloc] initWithAmount:amount
                                                            order:order
                                                         currency:product.currency
                                               productDescription:nil
                                                            owner:nil
                                                          scoring:nil];
    PTPVUser *user = [[PTPVUser alloc] initWithIdUser:card.idUser tokenUser:card.tokenUser];

    // make the request to PAYTPV
    [[PTPVAPIClient sharedClient] executePurchase:purchaseRequest forUser:user completion:^(PTPVPurchase * _Nullable purchase, NSError * _Nullable error) {
        [loader dismissAnimated:true completion:nil];

        if (error != nil) {
            AlertBuilder *alert = [AlertBuilder new];
            [alert setMessage:error.localizedDescription];
            [alert addActionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
            [alert presentInController:self animated:true completion:nil];
            return;
        }

        // send the purchase details to the backend
        Purchase *localPurchase = [[Purchase alloc] initWithOrder:purchase.order
                                                         authCode:purchase.authCode
                                                         currency:purchase.currency
                                                             card:card];
        [ModelStore addPurchase:localPurchase completion:^{
            AlertBuilder *alert = [AlertBuilder new];
            [alert setMessage:@"Product purchased!"];
            [alert addActionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                [self.navigationController popToRootViewControllerAnimated:true];
            }];
            [alert presentInController:self animated:true completion:nil];
        }];
    }];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.products.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"ProductCell"];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleValue1 reuseIdentifier:@"ProductCell"];
    }

    Product *product = [self.products objectAtIndex:indexPath.row];

    cell.textLabel.text = product.name;
    cell.detailTextLabel.text = [NSString stringWithFormat:@"%@ %@", product.currency, product.price];

    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    self.selectedProduct = [self.products objectAtIndex:indexPath.row];

    CardsTableViewController *controller = [CardsTableViewController new];
    controller.delegate = self;

    [self.navigationController pushViewController:controller animated:true];
}

#pragma mark - CardsTableControllerDelegate

- (void)cardsTableController:(CardsTableViewController *)controller didSelectCard:(Card *)card {
    if (self.selectedProduct == nil) {
        return;
    }

    // ask for user's confirmation
    AlertBuilder *alert = [AlertBuilder new];
    [alert setMessage:[NSString stringWithFormat:@"Are you sure you want to purchase %@ for %@ %@",
                       self.selectedProduct.name,
                       self.selectedProduct.currency,
                       self.selectedProduct.price]];
    [alert addActionWithTitle:@"No" style:UIAlertActionStyleCancel handler:nil];
    [alert addActionWithTitle:@"Yes" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        // make the purchase
        [self purchaseProduct:self.selectedProduct withCard:card];
    }];
    [alert presentInController:self animated:true completion:nil];
}

@end

File: Example/PayTPV Swift Example/ProductsTableViewController.swift
//
//  ProductsTableViewController.swift
//  PAYTPV Swift Example
//
//  Created by Mihail Cristian Dumitru on 8/17/17.
//  Copyright © 2017 PAYTPV. All rights reserved.
//

import UIKit
import PAYTPV

class ProductsTableViewController: UITableViewController {

    // MARK: - Properties

    let products: [Product] = [
        Product(name: "Product A", price: 0.29, currency: PTPVCurrencyEUR),
        Product(name: "Product B", price: 1.99, currency: PTPVCurrencyEUR),
        Product(name: "Product C", price: 2.99, currency: PTPVCurrencyEUR),
        Product(name: "Product D", price: 3.99, currency: PTPVCurrencyEUR),
    ]

    var selectedProduct: Product?

    // MARK: - Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()

        self.setup()
    }

    // MARK: - Setup

    func setup() {
        self.title = "Products"

        self.navigationItem.leftBarButtonItem = UIBarButtonItem(title: "Purchases", style: .plain, target: self, action: #selector(self.onPurchasesButtonPressed))
    }

    // MARK: - Helpers

    func onPurchasesButtonPressed() {
        self.navigationController?.pushViewController(PurchasesViewController(), animated: true)
    }

    func purchase(product: Product, withCard card: Card) {
        // add a loading indicator
        let loader = AlertBuilder()
            .set(title: "Executing purchase")
            .addLoadingIndicator()
            .present(sender: self)

        // the amount must be in integer format, so we have to convert it
        // e.g. 1.00 EURO = 100, 4.50 EUROS = 450...
        let amount = NSNumber(integerLiteral: Int((product.price * 100.0).rounded()))

        // generate unique order reference
        let order = String(format: "%.0f_%d", Date().timeIntervalSince1970, arc4random_uniform(1000))

        // create the purchase request
        let purchaseRequest = PTPVPurchaseRequest(amount: amount,
                                                  order: order,
                                                  currency: product.currency,
                                                  productDescription: nil,
                                                  owner: nil,
                                                  scoring: nil)

        let user = PTPVUser(idUser: card.id, tokenUser: card.token)

        // make the request to PAYTPV
        PTPVAPIClient.shared().executePurchase(purchaseRequest, for: user, completion: { (purchase, error) in
            loader.dismiss()

            guard let purchase = purchase else {
                if let error = error {
                    let _ = AlertBuilder()
                        .set(message: error.localizedDescription)
                        .add(actionWithTitle: "OK")
                        .present(sender: self)
                    return;
                }
                let _ = AlertBuilder()
                    .set(message: "There was an unexpected error. Please try again later.")
                    .add(actionWithTitle: "OK")
                    .present(sender: self)
                return;
            }

            // send the purchase details to the backend
            let localPurchase = Purchase(order: purchase.order ?? "",
                                         authCode: purchase.authCode ?? "",
                                         currency: purchase.currency ?? "",
                                         card: card)
            ModelStore.add(purchase: localPurchase, completion: {
                let _ = AlertBuilder()
                    .set(message: "Product purchased!")
                    .add(actionWithTitle: "OK", style: .default, handler: { (_) in
                        let _ = self.navigationController?.popToRootViewController(animated: true)
                    })
                    .present(sender: self)
            })
        })
    }

    // MARK: - Table view data source

    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.products.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "ProductCell") ?? UITableViewCell(style: .value1, reuseIdentifier: "ProductCell")

        let product = self.products[indexPath.row]

        cell.textLabel?.text = product.name
        cell.detailTextLabel?.text = "\(product.currency) \(product.price)"

        return cell
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        self.selectedProduct = self.products[indexPath.row]

        let cardsController = CardsTableViewController()
        cardsController.delegate = self

        self.navigationController?.pushViewController(cardsController, animated: true)
    }

}

// MARK: - CardsTableControllerDelegate

extension ProductsTableViewController: CardsTableControllerDelegate {

    func cardsTableController(_ controller: CardsTableViewController, didSelectCard card: Card) {
        guard let selectedProduct = self.selectedProduct else {
            return;
        }

        // ask for user's confirmation
        let _ = AlertBuilder()
            .set(message: "Are you sure you want to purchase \(selectedProduct.name) for \(selectedProduct.currency) \(selectedProduct.price)?")
            .add(actionWithTitle: "No", style: .cancel, handler: nil)
            .add(actionWithTitle: "Yes", style: .default) { (_) in
                // make the purchase
                self.purchase(product: selectedProduct, withCard: card)
            }
            .present(sender: self)
    }

}
